src/status_im2/contexts/chat/bottom_sheet_composer/utils.cljs

```diff
@@ -90,3 +93,57 @@
     (reanimated/set-shared-value last-height constants/input-height))
   (reset! text-value "")
   (rf/dispatch [:chat.ui/set-input-content-height constants/input-height]))
+
+(defn update-input
+  [{:keys [input-ref]}
+   {:keys [text-value]}
+   input-text]
+  (when (and input-text (not= @text-value input-text))
+    (reset! text-value input-text)
+    (when @input-ref
+      (.setNativeProps ^js @input-ref (clj->js {:text input-text})))))
+
+(defn count-lines
+  [s]
+  (-> s
+      (string/split #"\n" -1)
+      (butlast)
+      count))
+
+(defn cursor-y-position-relative-to-container
+  [{:keys [scroll-y]}
+   {:keys [cursor-position text-value]}]
+  (let [sub-text               (subs @text-value 0 @cursor-position)
+        sub-text-lines         (count-lines sub-text)
+        scrolled-lines         (Math/round (/ @scroll-y constants/line-height))
+        sub-text-lines-in-view (- sub-text-lines scrolled-lines)]
+    (* sub-text-lines-in-view constants/line-height)))
+
+(defn calc-suggestions-position
+  [cursor-pos {:keys [maximized?]} {:keys [height]} max-height count]
+  (let [{:keys [keyboard-height]} (hooks/use-keyboard)
+        insets                    (safe-area/get-insets)
+        curr-height               (reanimated/get-shared-value height)
+        window-height             (rf/sub [:dimensions/window-height])
```

Comment:
In my experience, we shouldn't mix code like this with re-frame and with hooks. `calc-suggestions-position` should be a pure function that receives all its arguments and then return a number. If it was a pure function it would be trivial to test it (as a unit test or experiment with it in the REPL).

To me, subscriptions and hooks should be reserved for building hiccup and nothing else. In `status-mobile` we make a common mistake of using subscriptions like function calls, that can be called anywhere, at any point in time. Let's just remember that calls to `rf/sub` are stateful, and the more stateful functions the more difficult to maintain the code (at least that's the Clojure mantra).

But practically speaking, there are many other places in the codebase doing something like this @OmarBasem. I think we now have enough Clojure maturity in the team to start to question this practice.

I'm cc'ing @flexsurfer to give some visibility to this matter.

Reply:
Alright I will refactor this function